src/core/pivotControls/AxisArrow.tsx

```diff
@@ -0,0 +1,154 @@
+import * as React from 'react'
+import * as THREE from 'three'
+import { useThree } from '@react-three/fiber'
+import { Line } from '../Line'
+import { context } from '.'
+
+const vec1 = new THREE.Vector3()
+const vec2 = new THREE.Vector3()
+
+export const calculateOffset = (
+  clickPoint: THREE.Vector3,
+  normal: THREE.Vector3,
+  rayStart: THREE.Vector3,
+  rayDir: THREE.Vector3
+) => {
+  const e1 = normal.dot(normal)
+  const e2 = normal.dot(clickPoint) - normal.dot(rayStart)
+  const e3 = normal.dot(rayDir)
+
+  vec1
+    .copy(rayDir)
+    .multiplyScalar(e1 / e3)
+    .sub(normal)
+  vec2
+    .copy(rayDir)
+    .multiplyScalar(e2 / e3)
+    .add(rayStart)
+    .sub(clickPoint)
+
+  const offset = -vec1.dot(vec2) / vec1.dot(vec1)
+  return offset
+}
+
+const upV = new THREE.Vector3(0, 1, 0)
+const offsetMatrix = new THREE.Matrix4()
+
+export const AxisArrow: React.FC<{ direction: THREE.Vector3; axis: 0 | 1 | 2 }> = ({ direction, axis }) => {
+  const {
+    depthTest,
+    scale,
+    lineWidth,
+    fixed,
+    axisColors,
+    hoveredColor,
+    opacity,
+    onDragStart,
+    onDrag,
+    onDragEnd,
+    userData,
+  } = React.useContext(context)
+
+  const camControls = useThree((state) => state.controls as any)
+  const objRef = React.useRef<THREE.Group>(null!)
+  const clickInfo = React.useRef<{ clickPoint: THREE.Vector3; dir: THREE.Vector3 } | null>(null)
+  const [isHovered, setIsHovered] = React.useState(false)
+
+  const onPointerDown = React.useCallback(
+    (e: any) => {
```

Comment:
This should be correctly typed.

Reply:
Also why are we wrapping these in `useCallback`?

Reply:
its problematic, i think fiber has wrong types. e.target is not defined by `ThreeEvent<PointerEvent>`. should we fix it here? `ThreeEvent<PointerEvent & { target: { capture..... }>`

Reply:
yeah, this is more complex than i thought. if i change the event type it wont fit into the fiber element props. the correct approach imo is `ThreeEvent<PointerEvent>` and a `// @ts-ignore` until fiber is fixed upstream.